man/vzfsync.8.in
.TH vzfsync 8 "31 Mar 2014" "OpenVZ" "Containers"
.SH NAME
vzfsync \- perform fsync/fdatasync/fadvise on files
.SH SYNOPSIS
.SY vzfsync
{
.BR -s , --sync
|
.BR -d , --datasync
}
.OP -n\fR,\fB--dontneed
.B file
.OP file ...
.YS
.SY vzfsync
.B --help
.YS
.SH DESCRIPTION
This utility performs
.BR fsync (2)
or
.BR fdatasync(2)
for every supplied file name.
It can also optionally perform
.BR posix_fadvise (2)
with the argument of
.BR POSIX_FADV_DONTNEED .
.P
It is used from within
.BR vzmigrate (8)
utility for live migration, right after copying ploop delta files
(and before suspending a container) to sync those copied files to disk
in order to optimize subsequent ploop mount time and, utlimately,
container frozen time.
.SH OPTIONS
.TP
.BR -s , --sync
Perform fsync.
.TP
.BR -d , --datasync
Perform fdatasync.
.TP
.BR -n , --dontneed
Perform posix_fadvise(POSIX_FADV_DONTNEED).
.SH EXIT STATUS
Returns 0 upon success, or an appropriate error code in case of an error:
.IP 1
Invalid usage
.IP 2
Failed to perform some requested operations on one or many files.
.SH SEE ALSO
.BR fsync (2),
.BR fdatasync (2),
.BR posix_fadvise (2).
.SH LICENSE
Copyright (C) 2014, Parallels, Inc. Licensed under GNU GPL v2.


ChangeLog
Since after 3.0.12 release, this file is obsoleted. Complete change log
can be obtained from OpenVZ git repository.

	http://git.openvz.org/?p=vzctl;a=shortlog
	http://git.openvz.org/?p=vzctl;a=log

Below is a change log for old vzctl releases.


   3.0.12: Thu Oct 5 2006 <igor@openvz.org>
	* vzctl start: added --wait cmdline option
	* added vznetcfg script for veth setup
	* fixed error message in case kernel compiled without USER_RESOURCES
	* added '--applyconfig_map name' vzctl set option
	* changed friendly name data structure
	* fixed vzmigrate to support VE name feature (bug #239)
	* fixed annoying arp warning (bug #189)
	* added bash completion script (bug #212)
	* added check to vzmigrate if OpenVZ is run on another side (bug #163)
	* removed debian directory from the tarball (bugs #197, #194)
	* removed LICVIEW from /etc/init.d/vz (bug #237)
	* fixed misprints and /dev/nul(l) usage in dist scripts (bug #228)
	* fixed vzmigrate to work with different awk flavors (bug #231)
	* years in copyright fixed (bug #232)
	* increased some UBCs in ve-light.conf-sample (bug #234)
	* fixed some minor memory leaks while doing realloc()
	* vzctl destroy: remove VE names on destroy (bug #236)
	* added --name option description to vzctl(8) man page (bug #232)
	* merged libvzctl-simfs and libvzctl into a single libvzctl
	* vzlist: replaced sprintf() with snprintf() to avoid buffer overflows
	* vzctl enter: forward error messages from child process
	* vzctl enter: fix off-by-one error which resulted in
	  uninitialized PATH
	* arpsend: fixed segfault when parsing long options
	* vzlist: fixed cast from pointer to integer of different size warnings
	* vzctl enter: exit with correct error code on enter fail (bug #245)
	* added udev rules for /dev/vzctl (bug #204)
	* added syscall numbers for powerpc
	* /etc/init.d/vz: check for /dev/vzctl only after modprobe (bug #204)
	* fixed ppc64 compilation
	* removed distro-specific stuff from 'make install' (bug #254)
	* added 'make install-gentoo' and Gentoo init.d/vz script (bug #262)
	* fixed arpsend options parsing for some compilers (bug #261)
	* man pages cleanup (VPS->VE, etc.)
	* changed all utilities and man pages to refer to VE not VPS
	* vzcpucheck: give error if VE0 CPU units not set (bug #249)

   3.0.11: Mon Aug  7 2006 <igor@openvz.org>
	* fixed using VE_ROUTE_SRC_DEV variable
	* added VPS name support
	* added change number of CPUs available in the VPS functionality.
	* fixes in CPT error processing
	* added upport of xt_mac iptables module
	* fixed awk expression to be compatible with mawk

   3.0.10: Wed Jun  7 2006 <igor@openvz.org>
	* added virtual ethernet device support (--veth_add, --veth_del)
	* added  /proc/meminfo virtualization support (--meminfo mode:value)
	* added VE0 UB setup functionality (--reset-ub)

   3.0.9: Wed May 24 2006 <igor@openvz.org>
	vzctl:
	* switched to new versioning scheme.
	* added -z option to /sbin/init
		(http://bugzilla.openvz.org/show_bug.cgi?id=164)
	* reset standart file desc [0-2] to /dev/null in case they are closed.
	* added possibility to set cpulimit in %
	* added suffixes G,M,K support for --diskspace option
	* set SIGPIPE handler to SIG_IGN before fist write to stdout/err
	* added Debian packaging support
	* changed directory structure.
		/etc/sysconfig/vz-scripts -> /etc/vz/conf
		/etc/sysconfig/vz-scripts/dists -> /etc/vz/dists
		/etc/sysconfig/vz -> /etc/vz/vz.conf
	* better distribution detection in vz initscript.
	* added Ubuntu distribution configuration file as copy of debian.
		(http://bugzilla.openvz.org/show_bug.cgi?id=119)
	* Increased kmemsize, privmpages, numfiles for x86_64
		ve-vps.basic.conf-sample.
		(http://bugzilla.openvz.org/show_bug.cgi?id=69)
	* changed -p -> --preserve=mode,ownership for 'cp' workaround in
		case /proc is not mounted.
	* internal modifications
		changed dynamic UB structure to static one.
		moved VPS configure functionality to vps_configure.c
		moved destroy functionaliny to destroy.c
	vzmigrate:
	* move config as veid.migrated to allow backward migration
		(http://bugzilla.openvz.org/show_bug.cgi?id=152)
	* added coping of action scripts.
		(http://bugzilla.openvz.org/show_bug.cgi?id=157)
	* fixed --keep-dst option processing. Option behavior was inverted.

   3.0.0-8: Mon Apr 24 2006 <igor@openvz.org>
	* UBC parameters > LONG_MAX in VPS conf truncate to LONG_MAX
	* added installation of vzmigrate.8 man page
	* use 'uname -m' for arch selection

   3.0.0-6: Tue Apr 18 2006 <kir@openvz.org>
	* man page fixes
   3.0.0-5: Fri Apr  7 2006 <igor@sw.ru>
	* added checkpointing support
	* added kernel headers for standalone compilation
	* chaged license from QPL to GPL
	* set personality PER_LINUX32 on x86_64 arch for i386 VPSes
	* fixed --iptables store (do not store dependent modules)
	* Do skip unknown iptables modules in config
	* set arg[0] = "-bash" on enter

   3.0.0-4: Thu Mar 16 2006 <kir@openvz.org>
	* cron scripts do nothing if VZ is not loaded (bug #107)
	* cosmetic vzctl.8 man page fixes
	* reverted bug #110 changes, see http://bugzilla.openvz.org/111
	* added ipt_REDIRECT support, thanks to jstubbs-at-gentoo.org

   3.0.0-3: Thu Mar  2 2006 <igor@sw.ru>
	* put_param3() can incorrectly set $name=( "" ) (bug #111).
		Affects gentoo VPSs. Thanks to jstubbs-at-gentoo.org
	* fixed compilation against old kernels (<= 022stab070)
	* description of set parameter suffixes (K, M, G, P) added
		to vzctl.8 man page (bug #105)
	* fixed gentoo-add_ip.sh script (bug #110)

   3.0.0-2: Thu Feb 23 2006 <igor@sw.ru>
	* fixed VPS start on old kernel (witthout feature_mask support)

   3.0.0-1: Wed Feb 22 2006 <igor@sw.ru>
	* enable sysfs for SuSE like distributions
	* added possibility to setup UBC parameters in the form
		xxx[GMKPB]:yyy[GMKPB]

   2.7.0-27: Tue Feb  7 2006 <igor@sw.ru>
	* fixed VE_CONFFILE variable passed to vps.mount script
	* fixed venet0 initialization in vz initscript (bug #99)

   2.7.0-26: Thu Feb  2 2006 <igor@sw.ru>

src/vzfsync.c
/*
 *  Copyright (C) 2014, Parallels, Inc. All rights reserved.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>

static int usage(int rc)
{
	FILE *out = (rc == 0)? stdout : stderr;

	fprintf(out,
"Usage: vzfsync {-s|-d} [-n] file [file ...]\n"
"       -s, --sync		do fsync()\n"
"       -d, --datasync		do fdatasync()\n"
"       -n, --dontneed		do fadvise(DONTNEED)\n"
	);

	return rc;
}

int main(int argc, char *argv[])
{
	int i;
	int fails = 0;
	int sync = 0, datasync = 0, dontneed = 0;
	struct option longopt[] = {
		{ "help",	no_argument, 0, 'h' },
		{ "sync",	no_argument, 0, 's' },
		{ "datasync",	no_argument, 0, 'd' },
		{ "dontneed",	no_argument, 0, 'n' },
		{}
	};

	while ((i = getopt_long(argc, argv, "hsdn", longopt, NULL)) != EOF) {
		switch (i) {
		   case 'h':
			return usage(0);
			break;
		   case 's':
			sync = 1;
			break;
		   case 'd':
			datasync = 1;
			break;
		   case 'n':
			dontneed = 1;
			break;
		   default:
			return usage(1);
		}
	}

	argc -= optind;
	argv += optind;

	if (argc < 1)
		return usage(0);

	if (sync == 1 && datasync == 1) {
		fprintf(stderr, "Options -s and -d can't be used together\n");
		return usage(1);
	}
	if (sync == 0 && datasync == 0 && dontneed == 0) {
		fprintf(stderr, "No options given\n");
		return usage(1);
	}

	/* Process files one by one */
	for (; argc > 0; argc--, argv++) {
		const char *name = argv[0];
		int fd;

		fd = open(name, O_RDONLY|O_DIRECT);
		if (fd < 0) {
			fprintf(stderr, "Can't open %s: %m\n", name);
			fails++;
			continue;
		}

		if (sync && fsync(fd) < 0) {
			fprintf(stderr, "Error fsync(%s): %m\n", name);
			fails++;
		}
		if (datasync && fdatasync(fd) < 0) {
			fprintf(stderr, "Error fdatasync(%s): %m\n", name);
			fails++;
		}
		if (dontneed && posix_fadvise(fd, 0, 0,
					POSIX_FADV_DONTNEED) < 0) {
			fprintf(stderr, "Error fadvise(%s): %m\n", name);
			fails++;
		}

		if (close(fd) < 0) {
			fprintf(stderr, "Error close(%s): %m\n", name);
			fails++;
		}
	}

	return (fails == 0) ? 0 : 2;
}
